src/tls/rustls.rs

```diff
@@ -121,13 +124,32 @@ where
 }
 
 #[async_trait::async_trait]
-impl<S> Reconnect for TlsStream<S>
-where
-    S: AsyncRead + AsyncWrite + Send + Unpin + Reconnect,
-{
-    async fn reconnect(&mut self) -> io::Result<Self> {
-        let stream = self.stream.get_mut().0.reconnect().await?;
-        TlsStream::new(stream, self.connector.clone(), self.hostname.clone()).await
+impl Reconnect for BufStream<TlsStream<tokio::net::TcpStream>> {
+    async fn reconnect(&mut self) -> io::Result<proto::BoxedConnection> {
+        let stream = self.get_mut().stream.get_mut().0.reconnect().await?;
+        let tls_stream = TlsStream::new(
+            stream,
+            self.get_ref().connector.clone(),
+            self.get_ref().hostname.clone(),
+        )
+        .await?;
+        let buffered = BufStream::new(tls_stream);
+        Ok(Box::new(buffered))
+    }
+}
+
+#[async_trait::async_trait]
+impl Reconnect for BufStream<TlsStream<proto::BoxedConnection>> {
```

Comment:
This `impl` can probably go away with the two blanket impls I've outlined above (at least assuming we `impl Reconnect for BoxedConnection`).